src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/BatchEmitter.java

```diff
@@ -216,10 +238,15 @@ private Runnable getPostRequestRunnable(int numberOfEvents) {
                 if (isSuccessfulSend(code)) {
                     LOGGER.debug("BatchEmitter successfully sent {} events: code: {}", eventsInRequest.size(), code);
                     retryDelay.set(0L);
-                    eventStore.cleanupAfterSendingAttempt(true, batchedEvents.getBatchId());
+                    eventStore.deleteBatchedEvents(true, batchedEvents.getBatchId());
                 } else {
-                    LOGGER.error("BatchEmitter failed to send {} events: code: {}", eventsInRequest.size(), code);
-                    eventStore.cleanupAfterSendingAttempt(false, batchedEvents.getBatchId());
+                    if (fatalResponseCodes.contains(code)) {
```

Comment:
I left it trying to send indefinitely with exponential backoff. But actually it might be better to just `close()` the Emitter

Reply:
IMO It's ok. I think there are different possible implementations. Hard to say which one can be the best for the different use cases.